File: daily_python/860_Lemonade_Change.py
import os
import sys
from typing import List

class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        five = 0
        ten = 0
        for i in bills:
            if i == 5:
                five += 1
            elif i == 10:
                if five == 0:
                    return False
                five -= 1
                ten += 1
            else:
                if ten > 0 and five > 0:
                    ten -= 1
                    five -= 1
                elif five > 2:
                    five -= 3
                else:
                    return False
        return True
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            bills = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(bills)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    solution = Solution()
    results = []
    for bills in dataset:
        results.append(solution.lemonadeChange(bills))
    for index, result in enumerate(results):
        print(f'Example {index+1} : {result}')

File: daily/2024/860_Lemonade_Change.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        int five = 0;
        int ten = 0;
        for (int i : bills){
            if (i == 5) five++;
            else if (i == 10){
                five--, ten++;
                if (five < 0) return false;
            }
            else{
                if (ten > 0 && five > 0) ten--, five--;
                else if (five > 2) five -= 3;
                else return false;
            }
        }
        return true;
    }
};

int main(){
    // vector<int> bills = {5,5,5,10,20};
    vector<int> bills = {5,5,10,10,20};
    Solution S;
    cout << S.lemonadeChange(bills) << endl;
    return 0;
}